examples/tutorials/mocking.md

```diff
@@ -0,0 +1,594 @@
+testing library:
+
+```ts
+import { assertEquals } from "jsr:@std/assert";
+import { Stub, stub } from "jsr:@std/testing/mock";
+
+// Define types for better code quality
+interface User {
+  name: string;
+  role: string;
+}
+
+// Original function
+function getCurrentUser(userId: string): User {
+  // Implementation that might involve database calls
+  return { name: "Real User", role: "admin" };
+}
+
+// Function we want to test
+function hasAdminAccess(userId: string): boolean {
+  const user = getCurrentUser(userId);
+  return user.role === "admin";
+}
+
+Deno.test("hasAdminAccess with stubbed user", () => {
+  // Create a stub that replaces getCurrentUser
+  const getUserStub: Stub<typeof getCurrentUser> = stub(
+    globalThis,
+    "getCurrentUser",
+    // Return a test user with non-admin role
+    () => ({ name: "Test User", role: "guest" }),
+  );
+
+  try {
+    // Test with the stubbed function
+    const result = hasAdminAccess("user123");
+    assertEquals(result, false);
+
+    // You can also change the stub's behavior during the test
+    getUserStub.restore(); // Remove first stub
+
+    const adminStub = stub(
+      globalThis,
+      "getCurrentUser",
+      () => ({ name: "Admin User", role: "admin" }),
+    );
+
+    try {
+      const adminResult = hasAdminAccess("admin456");
+      assertEquals(adminResult, true);
+    } finally {
+      adminStub.restore();
+    }
+  } finally {
+    // Always restore the original function
+    getUserStub.restore();
+  }
+});
+```
+
+Here we import the necessary functions from the Deno Standard Library, then we
+set up the function we're going to stub. In a real application this might
+connect to a database, make an API call, or perform other operations that we may
+want to avoid during testing.
+
+We set up the function under test, in this case the `hasAdminAccess()` function.
+We want to test whether it:
+
+- Calls the `getCurrentUser()` function to get a user object
+- Checks if the user's role is "admin"
+- Returns a boolean indicating whether the user has admin access
+
+Next we create a test named `hasAdminAccess with a stubbed user` and set up a
+stub for the `getCurrentUser` function. This will replace the real
+implementation with one that returns a user with a `guest` role.
+
+We run the test with the stubbed function, it will call `hasAdminAccess` with a
+user ID. Even though the real function would return a user with `admin` role,
+our stub returns `guest`, so we can assert that `hasAdminAccess` returns `false`
+(since our stub returns a non-admin user).
+
+We can change the stub behavior to return `admin` instead and assert that the
+function now returns `true`.
+
+At the end we use a `finally` block to ensure the original function is restored
+so that we don't accidentally affect other tests.
+
+### Stubbing environment variables
+
+For deterministic testing, you often need to control environment variables.
+Deno's Standard Library provides utilities to achieve this:
```

Comment:
both this code block, and the code clock below are quite long. I didn't spot that on each occasion you unpacked and explained them in detail underneath, so I spent time parsing them myself before discovering that helpful description underneath.

I suggest leaving a small hint that this is coming up so that the reader knows that they can jump back and forth between the code and your breakdown of it. Perhaps something like:

```suggestion
Deno's Standard Library provides utilities to achieve this as we see in this code, and the explanation of it which follows below
```